lib/web_ui/lib/src/engine/ulps.dart

```diff
@@ -0,0 +1,194 @@
+/// other.
+int floatFromListAs2sCompliment(Float32List source, int index) =>
+    signBitTo2sCompliment(float2Bits(source, index));
+
+int floatAs2sCompliment(double x) =>
+    signBitTo2sCompliment(_floatBitConverter.toBits(x));
+
+double twosComplimentAsFloat(int x) =>
+    bitsToFloat(twosComplimentToSignBit(x));
+
+bool _argumentsDenormalized(double a, double b, int epsilon) {
+  double denormalizedCheck = kFltEpsilon * epsilon / 2;
+  return a.abs() <= denormalizedCheck && b.abs() <= denormalizedCheck;
+}
+
+bool equalUlps(double a, double b, int epsilon, int depsilon) {
+  if (_argumentsDenormalized(a, b, depsilon)) {
+    return true;
+  }
+  int aBits = floatAs2sCompliment(a);
+  int bBits = floatAs2sCompliment(b);
+  // Find the difference in ULPs.
+  return aBits < bBits + epsilon && bBits < aBits + epsilon;
+}
+
+bool almostEqualUlps(double a, double b) {
+  const int kUlpsEpsilon = 16;
+  return equalUlps(a, b, kUlpsEpsilon, kUlpsEpsilon);
+}
+
+// Equality using the same error term as between
+bool almostBequalUlps(double a, double b) {
+  const int kUlpsEpsilon = 2;
+  return equalUlps(a, b, kUlpsEpsilon, kUlpsEpsilon);
+}
+
+bool almostPequalUlps(double a, double b) {
+  const int kUlpsEpsilon = 8;
+  return equalUlps(a, b, kUlpsEpsilon, kUlpsEpsilon);
+}
+
+bool almostDequalUlps(double a, double b) {
+  const int kUlpsEpsilon = 16;
+  return equalUlps(a, b, kUlpsEpsilon, kUlpsEpsilon);
+}
+
+bool approximatelyEqual(double ax, double ay, double bx, double by) {
+  if (approximatelyEqualT(ax, bx) && approximatelyEqualT(ay, by)) {
+    return true;
+  }
+  if (!roughlyEqualUlps(ax, bx) || !roughlyEqualUlps(ay, by)) {
+    return false;
+  }
+  final double dx = (ax - bx);
+  final double dy = (ay - by);
+  double dist = math.sqrt(dx * dx + dy * dy);
+  double tiniest = math.min(math.min(math.min(ax, bx), ay), by);
+  double largest = math.max(math.max(math.max(ax, bx), ay), by);
+  largest = math.max(largest, -tiniest);
+  return almostDequalUlps(largest, largest + dist);
+}
+
+// Use this for comparing Ts in the range of 0 to 1. For general numbers (larger and smaller) use
+// AlmostEqualUlps instead.
+bool approximatelyEqualT(double t1, double t2) {
+  return approximatelyZero(t1 - t2);
+}
+
+bool approximatelyZero(double value) => value.abs() < kFltEpsilon;
+
+bool roughlyEqualUlps(double a, double b) {
+  const int UlpsEpsilon = 256;
+  const int DUlpsEpsilon = 1024;
+  return equalUlps(a, b, UlpsEpsilon, DUlpsEpsilon);
+}
+
+bool dEqualUlpsEpsilon(double a, double b, int epsilon) {
+  int aBits = floatAs2sCompliment(a);
+  int bBits = floatAs2sCompliment(b);
+  // Find the difference in ULPs.
+  return aBits < bBits + epsilon && bBits < aBits + epsilon;
+}
+
+bool almostDequalUlpsDouble(double a, double b) {
+  if (a.abs() < kScalarMax && b.abs() < kScalarMax) {
+    return almostDequalUlps(a, b);
+  }
+  return (a - b).abs() / math.max(a.abs(), b.abs()) < kFltEpsilon * 16;
```

Comment:
kFltEpsilon * 16 is already computed as `kFltEpsilonOrderableErr` below, it might be reused (not sure if the name of the epsilon makes sense in this application)

Reply:
Done. also added comments to both.